cmake/android/OpenCVDetectAndroidSDK.cmake

```diff
@@ -0,0 +1,210 @@
+        else()
+          message(FATAL_ERROR "Android SDK Build Tools: autodetection failed. "
+                              "Specify ANDROID_SDK_BUILD_TOOLS_VERSION / ANDROID_SDK_BUILD_TOOLS_SUBDIR / ANDROID_SDK_BUILD_TOOLS variable to bypass autodetection.\n"
+                              "${__msg_BUILD_ANDROID_PROJECTS}")
+        endif()
+      endif()
+    else()
+      ocv_parse_properties_file("${ANDROID_SDK_BUILD_TOOLS}/source.properties"
+          ANDROID_BUILD_TOOLS
+          MSG_PREFIX "Android SDK Tools: "
+      )
+      if(NOT DEFINED ANDROID_BUILD_TOOLS_Pkg_Revision)
+        message(FATAL_ERROR "Android SDK Build Tools: Can't detect version: ANDROID_SDK_BUILD_TOOLS=${ANDROID_SDK_BUILD_TOOLS}\n"
+                            "Specify ANDROID_SDK_BUILD_TOOLS_VERSION variable to bypass autodetection.\n"
+                            "${__msg_BUILD_ANDROID_PROJECTS}")
+      else()
+        set(ANDROID_SDK_BUILD_TOOLS_VERSION "${ANDROID_BUILD_TOOLS_Pkg_Revision}" CACHE INTERNAL "Android SDK Build Tools version")
+        message(STATUS "Android SDK Build Tools: ver. ${ANDROID_SDK_BUILD_TOOLS_VERSION} (ANDROID_SDK_BUILD_TOOLS=${ANDROID_SDK_BUILD_TOOLS})")
+      endif()
+    endif()  # ANDROID_SDK_BUILD_TOOLS
+  endif()  # ANDROID_SDK_BUILD_TOOLS_VERSION
+endmacro()  # ocv_detect_android_sdk_build_tools
+
+
+if(BUILD_ANDROID_PROJECTS)
+  ocv_detect_android_sdk()
+  ocv_detect_android_sdk_tools()
+  ocv_detect_android_sdk_build_tools()
+
+  if(ANDROID_SDK_TOOLS_VERSION VERSION_LESS 14)
+    message(FATAL_ERROR "Android SDK Tools: OpenCV requires Android SDK Tools revision 14 or newer.\n"
+                        "${__msg_BUILD_ANDROID_PROJECTS}")
+  endif()
+
+  if(NOT ANDROID_SDK_TOOLS_VERSION VERSION_LESS 25.3.0)
+    message(STATUS "Android SDK Tools: Ant (Eclipse) builds are NOT supported by Android SDK")
+    ocv_update(ANDROID_PROJECTS_SUPPORT_ANT OFF)
+    if(NOT ANDROID_SDK_BUILD_TOOLS_VERSION VERSION_LESS 26.0.2)
+      # https://developer.android.com/studio/releases/gradle-plugin.html
+      message(STATUS "Android SDK Build Tools: Gradle 3.0.0+ builds support is available")
+      ocv_update(ANDROID_PROJECTS_SUPPORT_GRADLE ON)
+    endif()
+  else()
+    include(${CMAKE_CURRENT_LIST_DIR}/../OpenCVDetectApacheAnt.cmake)
+    if(ANT_EXECUTABLE AND NOT ANT_VERSION VERSION_LESS 1.7)
+      message(STATUS "Android SDK Tools: Ant (Eclipse) builds are supported")
+      ocv_update(ANDROID_PROJECTS_SUPPORT_ANT ON)
+    endif()
+  endif()
+
+  if(NOT DEFINED ANDROID_PROJECTS_BUILD_TYPE)
+    if(ANDROID_PROJECTS_SUPPORT_ANT)
+      ocv_update(ANDROID_PROJECTS_BUILD_TYPE "ANT")
+    elseif(ANDROID_PROJECTS_SUPPORT_GRADLE)
+      ocv_update(ANDROID_PROJECTS_BUILD_TYPE "GRADLE")
+    else()
+      message(FATAL_ERROR "Android SDK: Can't build Android projects as requested by BUILD_ANDROID_PROJECTS=ON variable.\n"
```

Comment:
I'd add "neither Ant nor Gradle has been found" comment to this error message to make it more clear.

Reply:
Gradle (wrapper-downloader) is usually shipped as part of Android project. So we don't check existence of this and message gradle not found is not correct (because we must ship it).
Here we check support from Android SDK at first.

Reply:
I've met this error and was confused at first. Then I looked through the code and found that I did not install _ant_ (I did not try to build with gradle). Other messages in this file provide more information regarding the root cause of the problem, but this one does not give any clues.